### DISCUSSION TITLE
Problems with hello_world.py

### DISCUSSION THREAD

**Discussion by flrnhbr1** [USC-ACTLab/crazyswarm#603] [Q&A] [answered]
Hi, I am currently starting to work with crazyswarm and crazyflies(2.1).

I installed and configured everything, but I have big troubles with the first hovering with HW.

Everytime I execute the hello_world.py, the crazyflie flips over or flies somewhere. 
I already tried to balance everything and also tried to start the drone manually (which is working completely fine)

I also tried to change the drone, but the issue remains.

Is it possible, that I made a mistake, when setting up crazyswarm?

Thank you in advance!

Regards Florian

**Comment by knmcguire** [comment]
Just to help out for the maintainers: it will be good to mention what kind of positioning system you are using or specifically which decks. If you are using one crazyflie, before flight, you can see in RVIZ if the position is converging and stable by just holding it in your hand. 

Also you should let us know on which version of the firmware of the crazyflie firmware you have flashed your crazyflie with.

**Comment by flrnhbr1** [comment]
Hello,

sorry you are right, I should have written more information.

I use Optitrack system to track the crazyflies with a single marker.

When I start RVIZ the drone position is stable, also when I move it, the position moves stable.
But in the hello_world.py, the data of the optitrack is not even used, so the motion Capture system should not influence the flight. (at least if I understand it right)

I flashed the drone in the crazyflie client with the latest FW (I think 2022.01)

**Reply by knmcguire** [reply]
Ah great thanks!

No the flight should not affect the tracking... unless the marker is under the propeller somehow or not visible. 

I will try the hover example tomorrow in our lab (we don't have optitrack but qualysis). But in the mean time I really advise you to have a good look at the[ hover_swarm.launch](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/launch/hover_swarm.launch) file and the [configuration instructions](https://crazyswarm.readthedocs.io/en/latest/configuration.html) if there isn't anything you are missing. For instance I see that the data streaming pane on motive needs to be changed, maybe that is not done correctly?

Also searching through the discussions you can double check if anybody else if having issues with this.

**Reply by knmcguire** [reply]
Maybe just double checking if changing the controller to 1 (PID) makes any difference as well. But I think that the mellinger default should easily handle take off with a new crazyflie.

**Reply by flrnhbr1** [reply]
Ok thank you very much for the quick answer in the first place!

I will have a look at this when I am in the lab again and also will try to change to PID. 

But regarding the setup of the configurations of the motion capture: 
In the script hello_world.py there are only two flight commands --> takeoff, and land.
So I never read or compute data from the Motion capture. Therefore this shouldn't affect it even if the configuration of, lets say the data streaming pane of motive, is false?
Or is the motion capture data used in the calculation of state estimation etc. even if i don't explicitly read it and compute it in the python script?


And yes it looks like there is another one with a similar problem, he quoted my question and posted it to "Issue"

**Reply by knmcguire** [reply]
As I understand correctly the motion capture data is all handled in the crazyswarm_server ROS node, which should run if the launch file is called. So it is correct that the python script in crazyswarm/scripts/ does not do anything with mocap, with the assumption that it all works correctly on the crazyswarm server side of things. 

This overview explains it quite well I think: https://crazyswarm.readthedocs.io/en/latest/overview.html

**Reply by flrnhbr1** [reply]
Ok thanks, I will have a look at the configuration and hey to change the Controller in the Lab then.

Thank you for the moment.

**Reply by knmcguire** [reply]
I tried crazyswarm out with the latest firmware in our lab, and it seems to work fine! So there is probably something else going on on your setup. Hope it will be indeed the radio firmware as @whoenig said!

**Comment by whoenig** [comment]
Perhaps you forgot to update your Crazyradio firmware? This needs to be done even if you use a "recently" bought radio. You can check by either installing an LED ring (should show green) or setting enable_pose_logging to true if the data is transmitted correctly.

**Reply by flrnhbr1** [reply]
Indeed I did not upgrade the crazyradios firmware. I will try asap.

**Comment by flrnhbr1** [comment, marked as answer]
I hav now tried to upadate the FW as @whoenig mentioned. Now the drone is flying a bit more stable, but the problem is still there.
I also tried to change the controller to PID as @knmcguire mentioned, but unfortunately this doesn't change anything.

Here i uploaded a video of the flight now: https://youtu.be/COTS-W2y-OM

I have also tried to changed the targetHeight to e.g. 0.5, but the same behavior appears.

**Reply by knmcguire** [reply]
That indeed looks pretty serious. That does not seem to be the cause of the controller indeed.

Could you share your hover_swarm.launch file perhaps? Then we can see exactly what parameters have been set, and then we have more information to help you.

**Reply by flrnhbr1** [reply]
```<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />

  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflieTypes.yaml" />
  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflies.yaml" />

  <node pkg="crazyswarm" type="crazyswarm_server" name="crazyswarm_server" output="screen" >
    <rosparam>
      # Logging configuration (Use enable_logging to actually enable logging)
      genericLogTopics: ["log1"]
      genericLogTopicFrequencies: [10]
      genericLogTopic_log1_Variables: ["stateEstimate.x", "stateEstimate.y", "stateEstimate.z"]
      # firmware parameters for all drones (use crazyflieTypes.yaml to set per type, or
      # allCrazyflies.yaml to set per drone)
      firmwareParams:
        commander:
          enHighLevel: 1
        stabilizer:
          estimator: 2 # 1: complementary, 2: kalman
          controller: 2 # 1: PID, 2: mellinger
        ring:
          effect: 16 # 6: double spinner, 7: solid color, 16: packetRate
          solidBlue: 255 # if set to solid color
          solidGreen: 0 # if set to solid color
          solidRed: 0 # if set to solid color
          headlightEnable: 0
        locSrv:
          extPosStdDev: 1e-3
          extQuatStdDev: 0.5e-1
        kalman:
          resetEstimation: 1
      # tracking
      motion_capture_type: "optitrack" # one of none,vicon,optitrack,qualisys,vrpn
      object_tracking_type: "libobjecttracker" # one of motionCapture,libobjecttracker
      send_position_only: False # set to False to send position+orientation; set to True to send position only
      motion_capture_host_name: "143.205.143.247"
      save_point_clouds: "/dev/null" # set to a valid path to log mocap point cloud binary file.
      print_latency: False
      write_csvs: False
      force_no_cache: False
      enable_parameters: True
      enable_logging: True
      enable_logging_pose: True
    </rosparam>
  </node>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
    <param name="dev" value="$(arg joy_dev)" />
  </node>

  <node pkg="crazyswarm" type="crazyswarm_teleop" name="crazyswarm_teleop" output="screen">
    <param name="csv_file" value="$(find crazyswarm)/launch/figure8_smooth.csv" />
    <param name="timescale" value="0.8" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyswarm)/launch/test.rviz"/>

  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_x" args="/cf2/log1/values[0]"/> -->
  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_roll" args="/cf1/log1/values[2] /cf1/log1/values[3]"/> -->

</launch>

**Reply by flrnhbr1** [reply]
I only changed:
motion_capture_type: "optitrack" # one of none,vicon,optitrack,qualisys,vrpn
object_tracking_type: "libobjecttracker" # one of motionCapture,libobjecttracker
motion_capture_host_name: "143.205.143.247"

And yesterday i tried out to log a few values of the crazyflie to see if there is something strange going on.
Therefore  enable_logging is on True and genericLogTopic_log1_Variables are set to a few logging variables

**Reply by flrnhbr1** [reply]
And i saw that you suggseted in #610 to take also a look in the mocap_helper.launch file? Did i understood it right, that there should be written the mocap settings, that i changed in hover_swarm.launch? 

Because in my mocap_helper.launch are the default settings ("vicon" as motion_capture_type and so on) and not the one i wrote to hover_swarm.launch.

**Reply by knmcguire** [reply]
Thanks! 
So from the video I see that you only use one marker. So the fact that you selected libobjecttracker is indeed correct. From the crazyswarm documentation, there seems to be quite some extra instructions to get this to work.
_"If using libobjecttracker as object_tracking_type disable all assets and make sure that labeled or unlabeled markers are being streamed."_
That is something that you could check with the mocap_helper.launch and I think that you can change it to optitrack and just check if there are no bodies being streamed.

Another thing that could also be a potential issue, is the carpet you are flying on. As the altitude is being handled from the IMU alone, if the carpet is generating a lot of static it can affect the IMU before take off. Although it does not seem like a carpet that would be very static... but you never know!.

Finally, I noticed this ticket in the issue list that might be interesting to you: https://github.com/USC-ACTLab/crazyswarm/issues/580. Here it states that when using single marker tracking, you need to put the `send_position_only` flag to `True`.  This is not very clear immediately clear from the documentation, but I see that in your launch file that is set on `False`. Perhaps that would be the quickest would be to try out to set it to `True`

**Reply by flrnhbr1** [reply]
Ok thanks.
I will try to change this flag today, and if this doesn't help I will try to start from a metalplate or something, to get rid of possible electrostatics.

**Reply by flrnhbr1** [reply]
It actually worked, thank you very much.

The problem was, as it seems, the electrostatic field of the carpet. I tried to start the drone from a ESD-Safe plate and it started without problems.
There you can see my first ever succesful flight with the crazyflie and crazyswarm :)
https://youtu.be/2q0vO0YEHig

**Reply by knmcguire** [reply]
Awesome! Glad that it worked now

So it wasn't necessary to set the send_position_only to True? So you confirmed it was only the static that was the problem?

**Reply by flrnhbr1** [reply]
I first tried to change the flag to True, but this didn't change anything. Nevertheless the drone flew away somewhere.  
And afterwards i tried the plate underneath and it worked.

Therefore I am pretty sure, that this was the problem.

**Reply by jpreiss** [reply]
The flag indeed should be set to True. But single-marker tracking depends much more on the IMU than full pose tracking, so it's not surprising that a problem with the IMU would ruin a single-marker flight. What was the flag value in your successful flight?

I believe setting the flag to False should cause a crash because the EKF is tuned (by default) to trust the attitude measurement more than the IMU. The PC will be sending identity quaternions for the attitude, so the CF will think it's hovering when it is flying off.

**Reply by flrnhbr1** [reply]
At the succesful flight the flag was True and i will let it be like that, because it was already mentioned it should be True with Single Marker.
